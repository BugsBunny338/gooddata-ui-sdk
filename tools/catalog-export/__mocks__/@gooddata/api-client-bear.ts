// (C) 2007-2018 GoodData Corporation
import catalogItemsResponse from "./loadItems.json";
import getObjectsResponse from "./getObjects.json";
import getObjectsByQueryResponse from "./getObjectsByQuery.json";
import getVisualizationsResponse from "./getVisualizations.json";

export default {
    catalogue: {
        loadItems() {
            return new Promise((resolve) => resolve(catalogItemsResponse));
        },
    },
    md: {
        getObjectsByQuery() {
            return new Promise((resolve) => resolve(getObjectsByQueryResponse));
        },
        getObjects() {
            return new Promise((resolve) => resolve(getObjectsResponse));
        },
        getVisualizations() {
            return new Promise((resolve) => resolve(getVisualizationsResponse));
        },
    },
};
