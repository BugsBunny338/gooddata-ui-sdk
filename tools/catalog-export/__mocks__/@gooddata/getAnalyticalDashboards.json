[
    {
        "author": "/gdc/account/profile/187aa70b3f3ed50ba4940e7b1966367d",
        "category": "analyticalDashboard",
        "contributor": "/gdc/account/profile/2e5a40d590f955c88251a8aead5267bd",
        "created": "2017-01-26 17:53:13",
        "deprecated": "0",
        "identifier": "afMA17GSbk31",
        "isProduction": 1,
        "summary": "",
        "tags": "",
        "title": "KPIs",
        "updated": "2020-11-18 22:14:11",
        "uri": "/gdc/md/mbuumy476p78ybcceiru61hcyr8i8lo8/obj/2663"
    },
    {
        "author": "/gdc/account/profile/187aa70b3f3ed50ba4940e7b1966367d",
        "category": "analyticalDashboard",
        "contributor": "/gdc/account/profile/187aa70b3f3ed50ba4940e7b1966367d",
        "created": "2017-05-03 15:03:50",
        "deprecated": "0",
        "identifier": "abBJlnxrfEWH",
        "isProduction": 1,
        "summary": "",
        "tags": "",
        "title": "Store management KPIs",
        "updated": "2018-01-23 04:43:37",
        "uri": "/gdc/md/mbuumy476p78ybcceiru61hcyr8i8lo8/obj/6853"
    },
    {
        "author": "/gdc/account/profile/187aa70b3f3ed50ba4940e7b1966367d",
        "category": "analyticalDashboard",
        "contributor": "/gdc/account/profile/187aa70b3f3ed50ba4940e7b1966367d",
        "created": "2018-11-12 11:29:01",
        "deprecated": "0",
        "identifier": "aahlrtUEgKkq",
        "isProduction": 1,
        "summary": "",
        "tags": "",
        "title": "Untitled",
        "updated": "2018-11-12 11:29:02",
        "uri": "/gdc/md/mbuumy476p78ybcceiru61hcyr8i8lo8/obj/8867"
    }
]
