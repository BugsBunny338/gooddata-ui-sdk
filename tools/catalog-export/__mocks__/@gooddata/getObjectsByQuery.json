[
    {
        "dataSet": {
            "content": {
                "urn": "urn:gooddata:date",
                "mode": "",
                "facts": [],
                "titleSuffix": "Timeline",
                "dataLoadingColumns": [],
                "attributes": [
                    "/gdc/md/xxoajsqvo85zw3t786qljpxd0xkiena2/obj/439",
                    "/gdc/md/xxoajsqvo85zw3t786qljpxd0xkiena2/obj/440",
                    "/gdc/md/xxoajsqvo85zw3t786qljpxd0xkiena2/obj/441",
                    "/gdc/md/xxoajsqvo85zw3t786qljpxd0xkiena2/obj/442",
                    "/gdc/md/xxoajsqvo85zw3t786qljpxd0xkiena2/obj/443",
                    "/gdc/md/xxoajsqvo85zw3t786qljpxd0xkiena2/obj/444",
                    "/gdc/md/xxoajsqvo85zw3t786qljpxd0xkiena2/obj/445",
                    "/gdc/md/xxoajsqvo85zw3t786qljpxd0xkiena2/obj/446",
                    "/gdc/md/xxoajsqvo85zw3t786qljpxd0xkiena2/obj/447",
                    "/gdc/md/xxoajsqvo85zw3t786qljpxd0xkiena2/obj/448",
                    "/gdc/md/xxoajsqvo85zw3t786qljpxd0xkiena2/obj/449",
                    "/gdc/md/xxoajsqvo85zw3t786qljpxd0xkiena2/obj/450",
                    "/gdc/md/xxoajsqvo85zw3t786qljpxd0xkiena2/obj/451",
                    "/gdc/md/xxoajsqvo85zw3t786qljpxd0xkiena2/obj/452",
                    "/gdc/md/xxoajsqvo85zw3t786qljpxd0xkiena2/obj/453",
                    "/gdc/md/xxoajsqvo85zw3t786qljpxd0xkiena2/obj/454",
                    "/gdc/md/xxoajsqvo85zw3t786qljpxd0xkiena2/obj/455",
                    "/gdc/md/xxoajsqvo85zw3t786qljpxd0xkiena2/obj/456"
                ],
                "identifierPrefix": "timeline"
            },
            "links": {
                "dataUploads": "/gdc/md/xxoajsqvo85zw3t786qljpxd0xkiena2/data/uploads/499"
            },
            "meta": {
                "author": "/gdc/account/profile/8d381a33fff78cf82392c837246d362e",
                "uri": "/gdc/md/xxoajsqvo85zw3t786qljpxd0xkiena2/obj/499",
                "tags": "",
                "created": "2010-08-02 08:13:42",
                "identifier": "timeline.dataset.dt",
                "deprecated": "0",
                "summary": "DataSet Date",
                "isProduction": 1,
                "title": "Date (Timeline)",
                "category": "dataSet",
                "updated": "2018-05-15 16:18:26",
                "contributor": "/gdc/account/profile/9bf17dfc423644e9bf92925bab65a3e9"
            }
        }
    },
    {
        "dataSet": {
            "content": {
                "mode": "SLI",
                "facts": [
                    "/gdc/md/xxoajsqvo85zw3t786qljpxd0xkiena2/obj/706",
                    "/gdc/md/xxoajsqvo85zw3t786qljpxd0xkiena2/obj/707"
                ],
                "customUploadTimestamp": 1566890302945695,
                "dataLoadingColumns": [
                    "/gdc/md/xxoajsqvo85zw3t786qljpxd0xkiena2/obj/62",
                    "/gdc/md/xxoajsqvo85zw3t786qljpxd0xkiena2/obj/64",
                    "/gdc/md/xxoajsqvo85zw3t786qljpxd0xkiena2/obj/132",
                    "/gdc/md/xxoajsqvo85zw3t786qljpxd0xkiena2/obj/134",
                    "/gdc/md/xxoajsqvo85zw3t786qljpxd0xkiena2/obj/17",
                    "/gdc/md/xxoajsqvo85zw3t786qljpxd0xkiena2/obj/19",
                    "/gdc/md/xxoajsqvo85zw3t786qljpxd0xkiena2/obj/21",
                    "/gdc/md/xxoajsqvo85zw3t786qljpxd0xkiena2/obj/23",
                    "/gdc/md/xxoajsqvo85zw3t786qljpxd0xkiena2/obj/25",
                    "/gdc/md/xxoajsqvo85zw3t786qljpxd0xkiena2/obj/27",
                    "/gdc/md/xxoajsqvo85zw3t786qljpxd0xkiena2/obj/29",
                    "/gdc/md/xxoajsqvo85zw3t786qljpxd0xkiena2/obj/31",
                    "/gdc/md/xxoajsqvo85zw3t786qljpxd0xkiena2/obj/33",
                    "/gdc/md/xxoajsqvo85zw3t786qljpxd0xkiena2/obj/35",
                    "/gdc/md/xxoajsqvo85zw3t786qljpxd0xkiena2/obj/37"
                ],
                "attributes": [
                    "/gdc/md/xxoajsqvo85zw3t786qljpxd0xkiena2/obj/679",
                    "/gdc/md/xxoajsqvo85zw3t786qljpxd0xkiena2/obj/680",
                    "/gdc/md/xxoajsqvo85zw3t786qljpxd0xkiena2/obj/682"
                ]
            },
            "links": {
                "dataUploads": "/gdc/md/xxoajsqvo85zw3t786qljpxd0xkiena2/data/uploads/165"
            },
            "meta": {
                "author": "/gdc/account/profile/8d381a33fff78cf82392c837246d362e",
                "uri": "/gdc/md/xxoajsqvo85zw3t786qljpxd0xkiena2/obj/165",
                "tags": "",
                "created": "2018-04-27 11:13:04",
                "identifier": "dataset.salesdetailfact",
                "deprecated": "0",
                "summary": "",
                "isProduction": 1,
                "title": "Sales Detail Fact",
                "category": "dataSet",
                "updated": "2019-08-27 07:19:45",
                "contributor": "/gdc/account/profile/9bf17dfc423644e9bf92925bab65a3e9"
            }
        }
    }
]
