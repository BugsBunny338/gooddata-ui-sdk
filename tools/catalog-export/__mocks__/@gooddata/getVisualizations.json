[
    {
        "link": "/gdc/md/xxoajsqvo85zw3t786qljpxd0xkiena2/obj/962",
        "locked": 0,
        "author": "/gdc/account/profile/8d381a33fff78cf82392c837246d362e",
        "tags": "",
        "created": "2018-04-27 14:23:11",
        "deprecated": "0",
        "identifier": "abhJpedgcfU2",
        "summary": "",
        "isProduction": 1,
        "title": "Labor Costs vs Scheduled Costs",
        "category": "visualizationObject",
        "updated": "2018-05-15 16:18:26",
        "unlisted": 0,
        "contributor": "/gdc/account/profile/9bf17dfc423644e9bf92925bab65a3e9"
    },
    {
        "link": "/gdc/md/xxoajsqvo85zw3t786qljpxd0xkiena2/obj/964",
        "locked": 0,
        "author": "/gdc/account/profile/8d381a33fff78cf82392c837246d362e",
        "tags": "",
        "created": "2018-04-27 14:23:11",
        "deprecated": "0",
        "identifier": "acPWDTL2bJeX",
        "summary": "",
        "isProduction": 1,
        "title": "Total Sales by City",
        "category": "visualizationObject",
        "updated": "2018-05-15 16:18:26",
        "unlisted": 0,
        "contributor": "/gdc/account/profile/9bf17dfc423644e9bf92925bab65a3e9"
    },
    {
        "link": "/gdc/md/xxoajsqvo85zw3t786qljpxd0xkiena2/obj/965",
        "locked": 0,
        "author": "/gdc/account/profile/8d381a33fff78cf82392c837246d362e",
        "tags": "",
        "created": "2018-04-27 14:23:11",
        "deprecated": "0",
        "identifier": "aaBN5UG3dXu4",
        "summary": "",
        "isProduction": 1,
        "title": "$ Total Sales by Location",
        "category": "visualizationObject",
        "updated": "2018-05-15 16:18:26",
        "unlisted": 0,
        "contributor": "/gdc/account/profile/9bf17dfc423644e9bf92925bab65a3e9"
    }
]
