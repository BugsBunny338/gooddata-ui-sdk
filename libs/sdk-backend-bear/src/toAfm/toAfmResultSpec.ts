// (C) 2007-2018 GoodData Corporation
import compact = require("lodash/compact");
import isEmpty = require("lodash/isEmpty");
import { ExecuteAFM } from "@gooddata/gd-bear-model";
import { convertFilter } from "./FilterConverter";
import { convertMeasure } from "./MeasureConverter";
import {
    attributeLocalId,
    bucketItems,
    bucketsFindAttribute,
    dimensionTotals,
    IAttribute,
    isAttribute,
    ITotal,
    totalIsNative,
    IExecutionDefinition,
} from "@gooddata/sdk-model";

function convertAttribute(attribute: IAttribute, idx: number): ExecuteAFM.IAttribute {
    const alias = attribute.attribute.alias;
    const aliasProp = alias ? { alias } : {};
    return {
        displayForm: attribute.attribute.displayForm,
        localIdentifier: attribute.attribute.localIdentifier || `a${idx + 1}`,
        ...aliasProp,
    };
}

function convertAFM(def: IExecutionDefinition): ExecuteAFM.IAfm {
    const attributes: ExecuteAFM.IAttribute[] = def.attributes.map(convertAttribute);
    const attrProp = attributes.length ? { attributes } : {};

    const measures: ExecuteAFM.IMeasure[] = def.measures.map(convertMeasure);
    const measuresProp = measures.length ? { measures } : {};

    const filters: ExecuteAFM.CompatibilityFilter[] = def.filters
        ? compact(def.filters.map(convertFilter))
        : [];
    const filtersProp = filters.length ? { filters } : {};

    const nativeTotals = convertNativeTotals(def);
    const nativeTotalsProp = nativeTotals.length ? { nativeTotals } : {};

    return {
        ...measuresProp,
        ...attrProp,
        ...filtersProp,
        ...nativeTotalsProp,
    };
}

function convertNativeTotals(def: IExecutionDefinition): ExecuteAFM.INativeTotalItem[] {
    // first find all native totals defined across dimensions
    const nativeTotals = def.dimensions
        .map(dimensionTotals)
        .reduce((acc, totals) => acc.concat(...totals), [])
        .filter(totalIsNative);

    return nativeTotals.map(t => {
        // then for each native total, look across buckets (if any) for an attribute that is specified in
        // the total definition
        const attribute = bucketsFindAttribute(def.buckets, t.attributeIdentifier);

        // well - there is native total for attribute that is nowhere in them buckets - this must be caught earlier
        if (!attribute) {
            // TODO: ensure totals referencing missing attributes is not possible
            throw new Error("invalid invariant");
        }

        // now, knowing the bucket and index of the attribute.. take all attributes that are before it in
        // the bucket
        const rollupAttributes = bucketItems(attribute.bucket)
            .slice(0, attribute.idx)
            .filter(isAttribute)
            .map(attributeLocalId);

        // and create native total such, that it rolls up all those attributes
        return {
            measureIdentifier: t.measureIdentifier,
            attributeIdentifiers: rollupAttributes,
        };
    });
}

function convertTotals(totals: ITotal[] = []): ExecuteAFM.ITotalItem[] {
    return totals.map(t => {
        const newItem: ExecuteAFM.ITotalItem = {
            type: t.type,
            attributeIdentifier: t.attributeIdentifier,
            measureIdentifier: t.measureIdentifier,
        };

        return newItem;
    });
}

function convertDimensions(def: IExecutionDefinition): ExecuteAFM.IDimension[] {
    return def.dimensions.map(dim => {
        const totals = convertTotals(dim.totals);
        const totalsProp = !isEmpty(totals) ? { totals } : {};

        const newDim: ExecuteAFM.IDimension = {
            itemIdentifiers: dim.itemIdentifiers,
            ...totalsProp,
        };

        return newDim;
    });
}

function convertResultSpec(def: IExecutionDefinition): ExecuteAFM.IResultSpec {
    // TODO: SDK8: why this???
    // Workaround because we can handle only 1 sort item for now
    const sortsProp = !isEmpty(def.sortBy) ? { sorts: def.sortBy.slice(0, 1) } : {};
    const dims = convertDimensions(def);
    const dimsProp = !isEmpty(dims) ? { dimensions: dims } : {};

    return {
        ...sortsProp,
        ...dimsProp,
    };
}

/**
 * Converts execution definition to AFM Execution
 *
 * @param def - execution definition
 * @returns AFM Execution
 */
export function toAfmExecution(def: IExecutionDefinition): ExecuteAFM.IExecution {
    return {
        execution: {
            afm: convertAFM(def),
            resultSpec: convertResultSpec(def),
        },
    };
}
