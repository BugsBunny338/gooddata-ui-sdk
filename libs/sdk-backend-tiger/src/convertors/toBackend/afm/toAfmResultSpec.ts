// (C) 2007-2020 GoodData Corporation
import compact = require("lodash/compact");
import isEmpty = require("lodash/isEmpty");
import { IExecutionDefinition } from "@gooddata/sdk-model";
import { ExecuteAFM } from "@gooddata/api-client-tiger";
import { convertVisualizationObjectFilter } from "./FilterConverter";
import { convertMeasure } from "./MeasureConverter";
import { convertAttribute } from "./AttributeConverter";
import { convertDimensions } from "./DimensionsConverter";

function convertAFM(def: IExecutionDefinition): ExecuteAFM.IAfm {
    const attributes: ExecuteAFM.IAttribute[] = def.attributes.map(convertAttribute);
    const attrProp = { attributes };

    const measures: ExecuteAFM.IMeasure[] = def.measures.map(convertMeasure);
    const measuresProp = { measures };

    const filters: ExecuteAFM.CompatibilityFilter[] = def.filters
        ? compact(def.filters.map(convertVisualizationObjectFilter))
        : [];
    const filtersProp = { filters };

    return {
        ...measuresProp,
        ...attrProp,
        ...filtersProp,
    };
}

function convertResultSpec(def: IExecutionDefinition): ExecuteAFM.IResultSpec {
    const convertedDimensions = convertDimensions(def);
    const dimsProp = !isEmpty(convertedDimensions) ? { dimensions: convertedDimensions } : {};

    return {
        ...dimsProp,
    };
}
/**
 * Converts execution definition to AFM Execution
 *
 * @param def - execution definition
 * @returns AFM Execution
 */
export function toAfmExecution(def: IExecutionDefinition): ExecuteAFM.IExecution {
    return {
        project: def.workspace,
        resultSpec: convertResultSpec(def),
        execution: {
            ...convertAFM(def),
        },
    };
}
