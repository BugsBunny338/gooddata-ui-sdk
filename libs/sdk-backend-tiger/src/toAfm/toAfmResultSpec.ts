// (C) 2007-2018 GoodData Corporation
import compact = require("lodash/compact");
import { ExecuteAFM } from "../gd-tiger-model/ExecuteAFM";
import { convertVisualizationObjectFilter } from "./FilterConverter";
import { convertMeasure } from "./MeasureConverter";
import {
    attributeId,
    attributesFind,
    bucketItems,
    bucketsFindAttribute,
    dimensionTotals,
    IAttribute,
    isAttribute,
    isUriQualifier,
    totalIsNative,
    MeasureGroupIdentifier,
    IExecutionDefinition
} from "@gooddata/sdk-model";
import { NotSupported } from "@gooddata/sdk-backend-spi";
import isEmpty = require("lodash/isEmpty");

function convertAttribute(attribute: IAttribute, idx: number): ExecuteAFM.IAttribute {
    const alias = attribute.attribute.alias;
    const aliasProp = alias ? { alias } : {};
    const displayFromQualifier = attribute.attribute.displayForm;

    if (isUriQualifier(displayFromQualifier)) {
        throw new NotSupported("Tiger backend does not allow specifying display forms by URI");
    }

    return {
        displayForm: displayFromQualifier,
        localIdentifier: attribute.attribute.localIdentifier || `a${idx + 1}`,
        ...aliasProp,
    };
}

function convertAFM(def: IExecutionDefinition): ExecuteAFM.IAfm {
    const attributes: ExecuteAFM.IAttribute[] = def.attributes.map(convertAttribute);
    const attrProp = attributes.length ? { attributes } : {};

    const measures: ExecuteAFM.IMeasure[] = def.measures.map(convertMeasure);
    const measuresProp = measures.length ? { measures } : {};

    const filters: ExecuteAFM.CompatibilityFilter[] = def.filters
        ? compact(def.filters.map(convertVisualizationObjectFilter))
        : [];
    const filtersProp = filters.length ? { filters } : {};

    const nativeTotals = convertNativeTotals(def);
    const nativeTotalsProp = nativeTotals.length ? { nativeTotals } : {};

    return {
        ...measuresProp,
        ...attrProp,
        ...filtersProp,
        ...nativeTotalsProp,
    };
}

function convertNativeTotals(def: IExecutionDefinition): ExecuteAFM.INativeTotalItem[] {
    // first find all native totals defined across dimensions
    const nativeTotals = def.dimensions
        .map(dimensionTotals)
        .reduce((acc, totals) => acc.concat(...totals), [])
        .filter(totalIsNative);

    return nativeTotals.map(t => {
        // then for each native total, look across buckets (if any) for an attribute that is specified in
        // the total definition
        const attribute = bucketsFindAttribute(def.buckets, t.attributeIdentifier);

        // well - there is native total for attribute that is nowhere in them buckets - this must be caught earlier
        if (!attribute) {
            // TODO: ensure totals referencing missing attributes is not possible
            throw new Error("invalid invariant");
        }

        // now, knowing the bucket and index of the attribute.. take all attributes that are before it in
        // the bucket
        const rollupAttributes = bucketItems(attribute.bucket)
            .slice(0, attribute.idx)
            .filter(isAttribute)
            .map(attributeId);

        // and create native total such, that it rolls up all those attributes
        return {
            measureIdentifier: t.measureIdentifier,
            attributeIdentifiers: rollupAttributes,
        };
    });
}

function convertDimensions(def: IExecutionDefinition): ExecuteAFM.IDimension[] {
    def.dimensions.forEach(dim => {
        dim.itemIdentifiers.forEach(item => {
            if (item === MeasureGroupIdentifier) {
                return;
            }

            const attr = attributesFind(def.attributes, item);

            if (!attr) {
                throw new Error(`invalid invariant: dimension specifies undefined attr ${item}`);
            }

            const attrQualifier = attr.attribute.displayForm;

            if (isUriQualifier(attrQualifier)) {
                throw new NotSupported("tiger does not support attributes specified by uri");
            }
        });

        if (dim.totals) {
            throw new NotSupported("Tiger backend does not support totals.");
        }
    });
    return def.dimensions;
}


function convertResultSpec(def: IExecutionDefinition): ExecuteAFM.IResultSpec {
    if (!isEmpty(def.sortBy)) {
        // tslint:disable
        console.warn("Tiger backend does not support sorting. Ignoring...");
    }

    const convertedDimensions = convertDimensions(def);
    const dimsProp = !isEmpty(convertedDimensions) ? { dimensions: convertedDimensions } : {};

    return {
        ...dimsProp,
    };
}

/**
 * Converts execution definition to AFM Execution
 *
 * @param def - execution definition
 * @returns AFM Execution
 */
export function toAfmExecution(def: IExecutionDefinition): ExecuteAFM.IExecution {
    return {
        project: def.workspace,
        resultSpec: convertResultSpec(def),
        execution: {
            ...convertAFM(def),
        },
    };
}
