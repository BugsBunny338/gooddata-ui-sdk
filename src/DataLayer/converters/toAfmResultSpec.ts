// (C) 2007-2018 GoodData Corporation
import compact = require("lodash/compact");
import flatMap = require("lodash/flatMap");
import get = require("lodash/get");
import { AFM, VisualizationObject } from "@gooddata/typings";
import { convertVisualizationObjectFilter } from "./FilterConverter";
import MeasureConverter from "./MeasureConverter";

function convertAttribute(
    attribute: VisualizationObject.IVisualizationAttribute,
    idx: number,
): AFM.IAttribute {
    const alias = attribute.visualizationAttribute.alias;
    const aliasProp = alias ? { alias } : {};
    return {
        displayForm: attribute.visualizationAttribute.displayForm,
        localIdentifier: attribute.visualizationAttribute.localIdentifier || `a${idx + 1}`,
        ...aliasProp,
    };
}

function convertAFM(visualizationObject: VisualizationObject.IVisualizationObjectContent): AFM.IAfm {
    const attributes: AFM.IAttribute[] = getAttributes(visualizationObject.buckets).map(convertAttribute);
    const attrProp = attributes.length ? { attributes } : {};

    const measures: AFM.IMeasure[] = getMeasures(visualizationObject.buckets).map(
        MeasureConverter.convertMeasure,
    );
    const measuresProp = measures.length ? { measures } : {};

    const filters: AFM.CompatibilityFilter[] = visualizationObject.filters
        ? compact(visualizationObject.filters.map(convertVisualizationObjectFilter))
        : [];
    const filtersProp = filters.length ? { filters } : {};

    const nativeTotals = convertNativeTotals(visualizationObject);
    const nativeTotalsProp = nativeTotals.length ? { nativeTotals } : {};

    return {
        ...measuresProp,
        ...attrProp,
        ...filtersProp,
        ...nativeTotalsProp,
    };
}

function getMeasures(buckets: VisualizationObject.IBucket[]): VisualizationObject.IMeasure[] {
    return buckets.reduce((result: VisualizationObject.IMeasure[], bucket: VisualizationObject.IBucket) => {
        const measureItems: VisualizationObject.IMeasure[] = bucket.items.filter(
            VisualizationObject.isMeasure,
        );

        return result.concat(measureItems);
    }, []);
}

function getNativeTotalAttributeIdentifiers(
    bucket: VisualizationObject.IBucket,
    total: VisualizationObject.IVisualizationTotal,
): string[] {
    const attributes = bucket.items.filter(VisualizationObject.isAttribute);

    const totalAttributeIndex = attributes.findIndex(
        attribute => attribute.visualizationAttribute.localIdentifier === total.attributeIdentifier,
    );

    return attributes
        .slice(0, totalAttributeIndex)
        .map(attribute => attribute.visualizationAttribute.localIdentifier);
}

function convertNativeTotals(
    visObj: VisualizationObject.IVisualizationObjectContent,
): AFM.INativeTotalItem[] {
    const nativeTotalsPerBucket = visObj.buckets.map(bucket => {
        const totals = bucket.totals || [];
        const nativeTotals = totals.filter(total => total.type === "nat");

        return nativeTotals.map(total => ({
            measureIdentifier: total.measureIdentifier,
            attributeIdentifiers: getNativeTotalAttributeIdentifiers(bucket, total),
        }));
    });

    return flatMap(nativeTotalsPerBucket);
}

function getAttributes(
    buckets: VisualizationObject.IBucket[],
): VisualizationObject.IVisualizationAttribute[] {
    return buckets.reduce(
        (result: VisualizationObject.IVisualizationAttribute[], bucket: VisualizationObject.IBucket) => {
            const items: VisualizationObject.IVisualizationAttribute[] = bucket.items.filter(
                VisualizationObject.isAttribute,
            );

            return result.concat(items);
        },
        [],
    );
}

function convertSorting(visObj: VisualizationObject.IVisualizationObjectContent): AFM.SortItem[] {
    if (visObj.properties) {
        let properties = {};
        try {
            properties = JSON.parse(visObj.properties);
        } catch {
            // tslint:disable-next-line:no-console
            console.error("Properties contains invalid JSON string.");
        }

        const sorts: AFM.SortItem[] = get(properties, "sortItems", []);
        return sorts ? sorts : [];
    }

    return [];
}

function convertResultSpec(visObj: VisualizationObject.IVisualizationObjectContent): AFM.IResultSpec {
    const sorts = convertSorting(visObj);
    // Workaround because we can handle only 1 sort item for now
    const sortsProp = sorts.length ? { sorts: sorts.slice(0, 1) } : {};

    return {
        ...sortsProp,
    };
}

export interface IConvertedAFM {
    afm: AFM.IAfm;
    resultSpec: AFM.IResultSpec;
}

/**
 * Converts visualizationObject to afm and resultSpec
 *
 * @method toAfmResultSpec
 * @param {VisualizationObject.IVisualizationObjectContent} visObj
 * @returns {IConvertedAFM}
 */
export function toAfmResultSpec(visObj: VisualizationObject.IVisualizationObjectContent): IConvertedAFM {
    const afm = convertAFM(visObj);
    return {
        afm,
        resultSpec: convertResultSpec(visObj),
    };
}
