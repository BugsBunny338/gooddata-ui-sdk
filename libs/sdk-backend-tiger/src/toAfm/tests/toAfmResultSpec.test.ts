// (C) 2020 GoodData Corporation
import { toAfmExecution } from "../toAfmResultSpec";
import { ReferenceLdm } from "@gooddata/reference-workspace";
import { defWithAlias, defWithoutFilters } from "./InvalidInputs.fixture";
import {
    emptyDef,
    newDefForItems,
    newPositiveAttributeFilter,
    defSetSorts,
    newAttributeSort,
    defSetDimensions,
    newDimension,
    newTotal,
    newTwoDimensional,
    MeasureGroupIdentifier,
} from "@gooddata/sdk-model";

const workspace = "test workspace";

describe("converts execution definition to AFM Execution", () => {
    const Scenarios: Array<[string, any]> = [
        ["empty definition", emptyDef(workspace)],
        ["definition that has one attribute with alias and one attribute without localId", defWithAlias],
        ["definition that has no filter", defWithoutFilters],
        [
            "definition that has filters",
            newDefForItems(
                workspace,
                [ReferenceLdm.Account.Name, ReferenceLdm.Activity.Subject, ReferenceLdm.Won],
                [newPositiveAttributeFilter(ReferenceLdm.Account.Name, ["myAccount"])],
            ),
        ],
        ["sorts", defSetSorts(emptyDef(workspace), [newAttributeSort(ReferenceLdm.Account.Name, "asc")])],
        ["dimensions", defSetDimensions(emptyDef(workspace), [newDimension(["localId1"])])],
    ];

    it.each(Scenarios)("should return AFM Execution with %s", (_desc, input) => {
        expect(toAfmExecution(input)).toMatchSnapshot();
    });

    it("throw when dimension has non-native totals", () => {
        const Total = newTotal("sum", ReferenceLdm.Won, ReferenceLdm.Account.Name);
        const Dimensions = newTwoDimensional(["localId1"], [MeasureGroupIdentifier, Total]);

        expect(() =>
            toAfmExecution(defSetDimensions(emptyDef(workspace), Dimensions)),
        ).toThrowErrorMatchingSnapshot();
    });
    it("throw error with dimensions with native totals but no attribute in bucket", () => {
        const Total = newTotal("nat", ReferenceLdm.Won, ReferenceLdm.Account.Name);
        const Dimensions = newTwoDimensional(["localId1"], [MeasureGroupIdentifier, Total]);

        expect(() =>
            toAfmExecution(defSetDimensions(emptyDef(workspace), Dimensions)),
        ).toThrowErrorMatchingSnapshot();
    });
});
