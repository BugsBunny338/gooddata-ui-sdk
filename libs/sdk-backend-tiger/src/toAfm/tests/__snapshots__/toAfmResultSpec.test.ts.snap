// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`converts execution definition to AFM Execution should return AFM Execution with definition that has filters 1`] = `
Object {
  "execution": Object {
    "attributes": Array [
      Object {
        "displayForm": Object {
          "identifier": Object {
            "id": "label.account.id.name",
            "type": "label",
          },
        },
        "localIdentifier": "a_label.account.id.name",
      },
      Object {
        "displayForm": Object {
          "identifier": Object {
            "id": "label.activity.id.subject",
            "type": "label",
          },
        },
        "localIdentifier": "a_label.activity.id.subject",
      },
    ],
    "filters": Array [
      Object {
        "positiveAttributeFilter": Object {
          "displayForm": Object {
            "identifier": Object {
              "id": "label.account.id.name",
              "type": "label",
            },
          },
          "in": Object {
            "values": Array [
              "myAccount",
            ],
          },
        },
      },
    ],
    "measures": Array [
      Object {
        "definition": Object {
          "measure": Object {
            "item": Object {
              "identifier": Object {
                "id": "acugFHNJgsBy",
                "type": "fact",
              },
            },
          },
        },
        "localIdentifier": "m_acugFHNJgsBy",
      },
    ],
  },
  "project": "test workspace",
  "resultSpec": Object {},
}
`;

exports[`converts execution definition to AFM Execution should return AFM Execution with definition that has no filter 1`] = `
Object {
  "execution": Object {
    "attributes": Array [],
    "filters": Array [],
    "measures": Array [],
  },
  "project": "test workspace",
  "resultSpec": Object {},
}
`;

exports[`converts execution definition to AFM Execution should return AFM Execution with definition that has one attribute with alias and one attribute without localId 1`] = `
Object {
  "execution": Object {
    "attributes": Array [
      Object {
        "alias": "alias",
        "displayForm": Object {
          "identifier": Object {
            "id": "attribute1",
            "type": "label",
          },
        },
        "localIdentifier": "a_724874d1_attribute1",
      },
      Object {
        "displayForm": Object {
          "identifier": Object {
            "id": "attribute2",
            "type": "label",
          },
        },
        "localIdentifier": "a_attribute2",
      },
    ],
    "filters": Array [],
    "measures": Array [],
  },
  "project": "test workspace",
  "resultSpec": Object {},
}
`;

exports[`converts execution definition to AFM Execution should return AFM Execution with dimensions 1`] = `
Object {
  "execution": Object {
    "attributes": Array [],
    "filters": Array [],
    "measures": Array [],
  },
  "project": "test workspace",
  "resultSpec": Object {
    "dimensions": Array [
      Object {
        "itemIdentifiers": Array [
          "localId1",
        ],
      },
    ],
  },
}
`;

exports[`converts execution definition to AFM Execution should return AFM Execution with empty definition 1`] = `
Object {
  "execution": Object {
    "attributes": Array [],
    "filters": Array [],
    "measures": Array [],
  },
  "project": "test workspace",
  "resultSpec": Object {},
}
`;

exports[`converts execution definition to AFM Execution should return AFM Execution with sorts 1`] = `
Object {
  "execution": Object {
    "attributes": Array [],
    "filters": Array [],
    "measures": Array [],
  },
  "project": "test workspace",
  "resultSpec": Object {},
}
`;

exports[`converts execution definition to AFM Execution throw error with dimensions with native totals but no attribute in bucket 1`] = `"Tiger backend does not support totals."`;

exports[`converts execution definition to AFM Execution throw when dimension has non-native totals 1`] = `"Tiger backend does not support totals."`;
